<a href="https://colab.research.google.com/github/JasperOwen/flyrank-ml-internship/blob/main/work/notebooks/w03_data_contract.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/JasperOwen/flyrank-ml-internship/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

**Unit of analysis.** Each row in the raw dataset represents one page's performance on one day, that belongs to one client. For the model input, I will format these daily records into summary records where each row represent's the performance across March 2026 of one page that belongs to one client, with columns that calculate perforamcne change such as drops in clicks


**Time window.** I will use March 2026 as my development month and 15th March 2026 as my decision point. This means that I will only use data from 1st - 15th March 2026 to identify signs of declining performance and create my refresh priority score. I will use data from after 15th March only to test if my model's rankings are correct.

**Table used.** I will use the fact_content_daily_performance table

**Exclusion.** The data from after 15th March 2026 will not be used when ranking the pages. This will be to prevent data leakage

**Label.** My label will be march_click_change.	This is a proxy for my refresh_priority_score in future capstone work. It will be used to predict the change in the number of clicks a page gets from the first half to the second half of march, with a decrease in clicks indicating the decline of a page.

In [1]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import os
from google.colab import userdata

hf_token = userdata.get("HF_TOKEN")
os.environ["HF_TOKEN"] = hf_token

import duckdb
from huggingface_hub import login

login(token=hf_token)

con = duckdb.connect()
con.sql("SET enable_http_metadata_cache=true;")

Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


In [2]:
from huggingface_hub import hf_hub_download

file_path = hf_hub_download(
    repo_id="FlyRank/internship-warehouse",
    repo_type="dataset",
    filename="fact_content_daily_performance/month=2026-03/data_0.parquet",
    token=hf_token
)

file_path

'/root/.cache/huggingface/hub/datasets--FlyRank--internship-warehouse/snapshots/50cbf7c3909d07be4d1b5906b4d09e882e5acbf2/fact_content_daily_performance/month=2026-03/data_0.parquet'

In [3]:
con.sql(f"""
    DESCRIBE SELECT *
    FROM read_parquet('{file_path}')
    LIMIT 1
""").df()

,column_name,column_type,null,key,default,extra
0,report_date,DATE,YES,None,None,None
1,client_hash_id,VARCHAR,YES,None,None,None
2,content_hash_id,VARCHAR,YES,None,None,None
3,client_has_gsc,BOOLEAN,YES,None,None,None
4,client_has_ga4,BOOLEAN,YES,None,None,None
5,gsc_data_available,BOOLEAN,YES,None,None,None
6,ga4_data_available,BOOLEAN,YES,None,None,None
7,gsc_impressions,BIGINT,YES,None,None,None
8,gsc_clicks,BIGINT,YES,None,None,None
9,gsc_sum_position,BIGINT,YES,None,None,None


## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

| Field | Bucket | Why |
| ------------------------------------------------------ | -------- | ------------------------------------ |
| `sessions_ai`                                           | Feature  | Shows how many users were directed to a page by AI, which indicates its visibility and how often AI cites it. Avaliable at the decision moment because AI referrals up to 15th March 2026 are recorded in the dataset before any decisions are made. |
| `gsc_clicks`                                      | Feature  | Shows how many users clicked on the page in a Google Search, making it a strong indicator of the page's performance. Avaliable at the decision moment because the number of clicks recorded up to 15th March 2026 is listed in the dataset. |
| `gsc_avg_position`                                     | Feature  | Shows how highly the page ranks in Google Search, which influences how likely users are to find it. Avaliable at the decision moment because the average positions up to 15th March 2026 are all included in the dataset.|
| `ga4_engaged_sessions`                                         | Feature  | Shows how many users actively engaged with the page, indicating the quality of its content. Avaliable at the decision moment because ga4 engagement data up to 15th March 2026 is recorded before any decision are made. |
| `ga4_total_engagement_sec`                                        | Feature  | Shows how long a user engaged with a page, which indicates how useful the content is. Avaliable at the decision moment because ga4 engagement data up to 15th March 2026 is recorded before any decision are made. |
| `march_click_change`                          | Label    | Used as a proxy to refresh_priority_score to indicate which pages are predicted to decline |
| `client_hash_id`                                       | Context  | Used to identify clients, not to be used when making predictions |
| `content_hash_id`                                      | Context  | Used to identify pages, not to be used when making predictions |
| `report_date`                                          | Context  | Used for aggregating data into monthly summaries only |
| `month`                                                | Context  | Allows me to organise the data by monthly performance and compare page performance across months |
| `ga4_data_available`                                   | Context  | Only indicates if a particular page has ga4 data avaliable, not relevent when making predictions |
| `client_has_gsc`                                       | Context  | Only indicates if a client has gsc data avaliable, not relevent when making predictions |
| `client_has_ga4`                                       | Context  | Only indicates if a client has ga4 data avaliable, not relevent when making predictions |
|                          |    |  ||
| Performance metrics past March 2026            | Excluded | These values are being used for testing, letting the model see them beforehand would be data leakage |
| `ga4_*` metrics where `ga4_data_available is not true` | Excluded | The data is unavaliable so it cannot be used |
| Rows where `gsc_data_available is not true`            | Excluded | The data is unavaliable so it cannot be used |

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

Each of three queries below verify the claims I have made above:

The first query is the grrain check. This confirms that my unit of analysis claim is true: that each row in the dataset is one page's performance in one day, where the page belongs to one client. By grouping the data by client_hash_id, content_hash_id, and report_date and finding no duplicates, I have shown that the grain holds

The second query confirms that the time window claim from seciton 1 is correct: that the development slice is March 2026. This is shown because the min and max dates outputted are March 1st 2026 and March 31st 2026

The third query is the avaliability check which confirms the avaliability of the data I will use with both gsc and ga4 data avaliable. This shows how much data my exclusion rules remove and how much data my model will be able to use.



In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
con.sql(f"""
SELECT
    client_hash_id,
    content_hash_id,
    report_date,
    COUNT(*) AS row_count
FROM read_parquet('{file_path}')
GROUP BY
    client_hash_id,
    content_hash_id,
    report_date
HAVING COUNT(*) > 1
""").df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,client_hash_id,content_hash_id,report_date,row_count


In [5]:
con.sql(f"""
SELECT
    COUNT(*)         AS slice_row_count,
    MIN(report_date) AS slice_min_date,
    MAX(report_date) AS slice_max_date
FROM read_parquet('{file_path}')
""").df()

,slice_row_count,slice_min_date,slice_max_date
0,9841378,2026-03-01,2026-03-31


In [6]:
con.sql(f"""
SELECT
    COUNT(*) AS usable_rows
FROM read_parquet('{file_path}')
WHERE
    gsc_data_available IS TRUE
    AND ga4_data_available IS TRUE
""").df()

,usable_rows
0,364347


In [7]:
feature_frame = con.sql(f"""
SELECT
    client_hash_id,
    content_hash_id,

    SUM(CASE WHEN report_date <= '2026-03-15' THEN sessions_ai ELSE 0 END) AS sessions_ai,

    SUM(CASE WHEN report_date <= '2026-03-15' THEN gsc_clicks ELSE 0 END) AS early_march_clicks,

    AVG(CASE WHEN report_date <= '2026-03-15' THEN gsc_avg_position ELSE NULL END) AS early_march_avg_position,

    SUM(CASE WHEN report_date <= '2026-03-15' THEN ga4_engaged_sessions ELSE 0 END) AS ga4_engaged_sessions,

    AVG(CASE WHEN report_date <= '2026-03-15' THEN ga4_total_engagement_sec ELSE 0 END) AS early_march_engagement_sec,

FROM read_parquet('{file_path}')
WHERE gsc_data_available IS TRUE
  AND ga4_data_available IS TRUE
GROUP BY
    client_hash_id,
    content_hash_id
""").df()

feature_frame["early_march_avg_position"] = feature_frame["early_march_avg_position"].fillna(999)

feature_frame.head()

,client_hash_id,content_hash_id,sessions_ai,early_march_clicks,early_march_avg_position,ga4_engaged_sessions,early_march_engagement_sec
0,client_9958f0a7ae1df715,content_810cf06597918291,0.0,0.0,8.795464,3.0,49.368421
1,client_9958f0a7ae1df715,content_b813c73d7000b3b1,0.0,1.0,7.290612,0.0,0.142857
2,client_9958f0a7ae1df715,content_5a77dbf5671c5a65,2.0,105.0,4.722870,9.0,18.967742
3,client_9958f0a7ae1df715,content_278030b007943b07,0.0,5.0,6.463128,1.0,8.900000
4,client_9958f0a7ae1df715,content_237e63fc00c8c7ad,0.0,15.0,6.955875,0.0,0.315789


### Leakage experiment


In [8]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
future = con.sql(f"""
SELECT
    client_hash_id,
    content_hash_id,

    SUM(CASE WHEN report_date > '2026-03-15' THEN gsc_clicks ELSE 0 END) AS late_march_clicks

FROM read_parquet('{file_path}')
WHERE gsc_data_available IS TRUE
  AND ga4_data_available IS TRUE
GROUP BY
    client_hash_id,
    content_hash_id
""").df()

future_frame = feature_frame.merge(
    future,
    on=["client_hash_id", "content_hash_id"],
    how="left"
)

In [9]:
future_frame['march_click_change'] = future_frame['late_march_clicks'] - future_frame['early_march_clicks']
future_frame.head(5)

,client_hash_id,content_hash_id,sessions_ai,early_march_clicks,early_march_avg_position,ga4_engaged_sessions,early_march_engagement_sec,late_march_clicks,march_click_change
0,client_9958f0a7ae1df715,content_810cf06597918291,0.0,0.0,8.795464,3.0,49.368421,1.0,1.0
1,client_9958f0a7ae1df715,content_b813c73d7000b3b1,0.0,1.0,7.290612,0.0,0.142857,0.0,-1.0
2,client_9958f0a7ae1df715,content_5a77dbf5671c5a65,2.0,105.0,4.722870,9.0,18.967742,94.0,-11.0
3,client_9958f0a7ae1df715,content_278030b007943b07,0.0,5.0,6.463128,1.0,8.900000,2.0,-3.0
4,client_9958f0a7ae1df715,content_237e63fc00c8c7ad,0.0,15.0,6.955875,0.0,0.315789,24.0,9.0


In [10]:
outcome_frame = future_frame[
    [
        "client_hash_id",
        "content_hash_id",
        "march_click_change"
    ]
].copy()

outcome_frame.head()

,client_hash_id,content_hash_id,march_click_change
0,client_9958f0a7ae1df715,content_810cf06597918291,1.0
1,client_9958f0a7ae1df715,content_b813c73d7000b3b1,-1.0
2,client_9958f0a7ae1df715,content_5a77dbf5671c5a65,-11.0
3,client_9958f0a7ae1df715,content_278030b007943b07,-3.0
4,client_9958f0a7ae1df715,content_237e63fc00c8c7ad,9.0


Shown below is a clean, leak-free attempt to predict the change in clicks from the first half to the second half of March 2026. Note the low R2 result, indicating that the leak-free model is guessing

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

clean_features = [
    "sessions_ai",
    "early_march_clicks",
    "early_march_avg_position",
    "ga4_engaged_sessions",
    "early_march_engagement_sec"
]

training_frame = feature_frame.merge(
    outcome_frame[
        [
            "client_hash_id",
            "content_hash_id",
            "march_click_change"
        ]
    ],
    on=[
        "client_hash_id",
        "content_hash_id"
    ]
)

X_clean = training_frame[clean_features]
y_clean = training_frame["march_click_change"]

X_train, X_test, y_train, y_test = train_test_split(
    X_clean,
    y_clean,
    test_size=0.2,
    random_state=42
)

clean_model = LinearRegression()

clean_model.fit(
    X_train,
    y_train
)

clean_predictions = clean_model.predict(X_test)

clean_score = r2_score(
    y_test,
    clean_predictions
)

clean_score

0.07862339935682128

Shown below is a leaky attempt to predict the change in clicks from the first half to the second half of March 2026. The leak is that the number of clicks from the second half of March 2026 is included with the input features. As a result the model gets a perfect R2 score of 1.0

In [12]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

leaky_features = [
    "sessions_ai",
    "early_march_clicks",
    "early_march_avg_position",
    "ga4_engaged_sessions",
    "early_march_engagement_sec",
    "late_march_clicks"
]

training_frame = future_frame.copy()

X_leaky = training_frame[leaky_features]
y_leaky = training_frame["march_click_change"]

X_train, X_test, y_train, y_test = train_test_split(
    X_leaky,
    y_leaky,
    test_size=0.2,
    random_state=42
)

leaky_model = LinearRegression()

leaky_model.fit(
    X_train,
    y_train
)

leaky_predictions = leaky_model.predict(X_test)

leaky_score = r2_score(
    y_test,
    leaky_predictions
)

leaky_score

1.0

## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

The dataset I will be using drops 96.3% of total dataset for March 2026 (364,347 rows used out of 9,841,378) due to using both gsc and ga4 data. The data can never show historical page performance for clients that do not use both gsc and ga4..

In addition, because the training data only uses March 2026, it is blind to long-term patterns. This means it will never be able to know if a decline in performance is true decline, seasonal decline or other pre-determined reasons that the page does not need refreshing for.

Also, the dataset can never note any changes made to the webpages during March 2026, it will only measure changes in features from the first and second halves of the month.

Finally, the data contains no marker for exactly which pages are in decline. Instead the label is a proxy score, meaning the data can never explicitly confirm that a page's performance is in decline.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.